### PR TITLE
Implement `java.util.PriorityQueue` from scratch.

### PR DESCRIPTION
**Pull request by sjrd**
~~Based on #3734. Only the last commit belongs to this PR.~~

Previously, it was implemented on top of `s.c.m.PriorityQueue`. Unlike other collections implemented on top of Scala collections, `PriorityQueue` had no serious hacks. It did have overhead in terms of allocations of `Box`es, though.

The new implementation, done from scratch, does not box elements, and offers direct support for all the methods of the JDK API.

The test suite of `PriorityQueue` is enhanced in the process. Notably, it includes all the standard tests for `Collection`s in `CollectionTest`.

### CODE REVIEW

File: javalib/src/main/scala/java/util/PriorityQueue.scala [hunk 1]
```diff
@@ -12,136 +12,181 @@
...
+  private[this] def fixUp(m: Int): Unit = {
+    val inner = this.inner // local copy
+    var k: Int = m
+    while (k > 1 && comp.compare(inner(k >> 1), inner(k)) > 0) {
+      val j = k >> 1
+      val h = inner(k)
+      inner(k) = inner(j)
+      inner(j) = h
+      k = j
+    }
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/PriorityQueue.scala]: Write this tailrec?

```scala
def fixUp(m: Int): Unit = {
  if (m > 1) {
    val parent = k >> 1
    if (comp.compare(inner(k >> 1), inner(k)) > 0) {
     // swap
     fixUp(parent)
    }
  }
}
```
Much easier to understand IMHO.

File: javalib/src/main/scala/java/util/PriorityQueue.scala [hunk 2]
```diff
@@ -12,136 +12,181 @@
...
   }
 
+  private[this] def fixDown(m: Int, n: Int): Unit = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/PriorityQueue.scala]: What is `m`, what is `n`?

Comment (gzm0) [reply javalib/src/main/scala/java/util/PriorityQueue.scala]: Why is `n` even a parameter? It's simply the ending index no? So we might as well take the length of `inner`?

File: javalib/src/main/scala/java/util/PriorityQueue.scala [hunk 3]
```diff
@@ -12,136 +12,181 @@
...
+      if (comp.compare(inner(k), inner(j)) <= 0) {
+        return
+      } else {
+        val h = inner(k)
+        inner(k) = inner(j)
+        inner(j) = h
+        k = j
+      }
+    }
+    // scalastyle:on return
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/PriorityQueue.scala]: Same here: Why not write this tailrec? I think it will be much easier to understand.

**Comment by sjrd** [comment]
Similarly to what I wrote in `TreeSet`, the names and algorithms have been kept the same as in the Scala version to preserve the similarity. I can change them if you think that intrinsic quality is worth more than the similarity.

**Comment by sjrd** [comment]
I have rebased this PR and made it independent of #3734.

I have added comments in the first commit; then I have added a second commit with tail-recursive implementations of `fixUp` and `fixDown`, so that you can compare which one is best.

**Comment by sjrd** [comment]
Well, I discovered several problems in the implementations of `PriorityQueue.remove(Object)` and `PriorityQueue#Iterator.remove()`, which were very subtle. But now I think I've nailed them. I have added test cases that are *very* whitebox, depending on speicific implementation details, because it was the only sure way to test certain corner cases.

**Comment by sjrd** [comment]
Given that the implementation had completely diverged from Scala's `PriorityQueue` by now, I have squashed the changes to tail-rec, and removed the reference to Scala's implementation in the commit message.

**Comment by gzm0** [comment]
Quick update here: The snapshoting behavior of the iterator kind of bothers me, but I can't come up with any better ideas.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala [hunk 4]
```diff
@@ -236,4 +241,192 @@ class PriorityQueueTest {
...
   }
 
+  @Test def poll(): Unit = {
+    val pq = newPriorityQueueWith0Until100()
+
+    var nextExpected = 0
+    while (!pq.isEmpty()) {
+      assertEquals(nextExpected, pq.poll())
+      nextExpected += 1
+    }
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala]: assert that nextExpected is 100?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala [hunk 5]
```diff
@@ -236,4 +241,192 @@ class PriorityQueueTest {
...
+        nextExpected += 1
+      }
+    }
+  }
+
+  @Test def removingAnArbitraryElementWithAnIteratorPreservesPrioritiesCornerCase(): Unit = {
+    /* This test depends on internal details of our implementation. It may have
+     * to be changed in the internal data structure and/or algorithms of
+     * PriorityQueue are modified in the future.
+     */
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala]: I'm not sure I understand why these tests must be whitebox tests in order to test what they test. Could you clarify this in the comments? (or if they don't have to, change them to blackbox tests)?

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/PriorityQueueTest.scala]: They don't need to be whitebox to test what they test. However, they need to be whitebox to *make sure* that they test what we think they're testing. In particular, the scenario where `Iterator.remove()` causes reordering in the heap such that elements yet to be iterated are moved before the cursor, and elements already iterated are moved after. This scenario is not easy to trigger. It requires some deliberate engineering of how the heap will look like and what specific element will be removed.

**Comment by sjrd** [comment]
Updated.
